agenthub/codeact_agent/codeact_agent.py

```diff
@@ -188,48 +207,62 @@ def step(self, state: State) -> Action:
         if latest_user_message and latest_user_message.strip() == '/exit':
             return AgentFinishAction()
 
-        # prepare what we want to send to the LLM
-        messages: list[dict[str, str]] = self._get_messages(state)
-
-        response = self.llm.completion(
-            messages=messages,
-            stop=[
-                '</execute_ipython>',
-                '</execute_bash>',
-                '</execute_browse>',
-            ],
-            temperature=0.0,
```

Comment:
and we will just have one additional argument here: `condense=True`?

Reply:
I did not get this. Why do we need to specify `condense=True`? 
Are you saying that if (`condense = False`) and context_window is exceeded, we don't condense the memory and just return an error?

Reply:
Yes! Basically, we want the user to use `llm.completion` with the default behavior of http://litellm.vercel.app/docs/completion. But we add ONE additional argument `condense`, when set to `True`, we automatically do all the fancy condensation stuff inside the `llm.completion` call, so user do not need be aware of what's happening under the hood, but just know "I can throw a gigantic list of messages into this API, and it will return the next action for me, and when i set `condense=True`, it can return the next action for me and costs me less!" 

Does this make sense?

Reply:
Okay. 
`LLM` class does not have access to the memory. To add the summary to state/history or event stream, I'll also have to pass that as an additional argument. 
Do you have any approach to add summary to history, without passing it as argument?

Reply:
If condensation is optional, then we could have `condense=True` as a configuration var, is there a reason not to? There are other memory-related parameters, and they're also configurable per app/llm/agent.

Reply:
@enyst hmm, good point - though it i think it will be nicer for agent developer to have more fine-grained control (e.g., only start condensing when it intended & do not condense requests that shouldn't be)

Reply:
@xingyaoww Well, it's possible for the agent developer to modify the llm config it sends, although I don't know, it might not be a great idea after the refactoring that removed individual params. It feels like it should be left to the user to set now. But there's nothing preventing modification of a setting, so if we want to, we can override it.

More importantly though, I'm not sure this boolean really has a use case for it, because this isn't quite accurate:
> "I can throw a gigantic list of messages into this API, and it will return the next action for me, and when i set condense=True, it can return the next action for me and costs me less!"

If it encounters a ContextWindowExceededError, then the task ends with an error, in case we don't try to condense. It doesn't return an action, no matter the cost, until the user restarts the task completely. Only if it condenses it can return an action at all. Moreover, if it continues it will cost more per task (since it didn't just stop after action N), but at least it works.

I wonder if what you may have had in mind actually is the token limit? If a token limit is set, then it will condense (much) sooner than the context limit. We have `max_input_tokens` for that, an already existing setting but unused in `main`, and this PR is using it: if set to something then it will try to respect that, condensing when it's larger, and of course, then it will cost less per task. By default it's None, so disabled.

Reply:
> I wonder if what you may have had in mind actually is the token limit?

Yep! You are right - if we add "condense=True" anywhere in the codebase, and the `max_input_tokens` is `None`, they will never be triggered, that is `condense=True` and `condense=False` have the same behavior, **until the chat exceed context window**. Since now we will error out when exceed context window, i think "not erroring out & run condensation" is already a step of improvement we can make :)